### PR TITLE
Softmax for find20

### PR DESCRIPTION
**Pull request by Vsevolod1983**
1. Softmax primitive is implemented in terms of Find 2.0 API 
2. GTest is added for testing Find 2.0 wrapper over Softmax for both SoftmaxForward and SoftmaxBackward

TICKETS: 
LWPMIOPEN-519
LWPMIOPEN-557

### CODE REVIEW

**Review by CAHEK7 — APPROVED** [review]

**Comment by Vsevolod1983** [comment]
Hi @junliume 
I can see "Read The Docs" is the only fail in my pull request. Clicking on Details brings me to 404 not found page.
Should I worry about this fail or its non-blocking thing ?

**Comment by junliume** [comment]
@atamazov @JehandadKhan please review since `miopen.h` is modified. Thanks!

Gentle ping to @DrizztDoUrden for find 2.0 review

**Comment by atamazov** [comment]
@Vsevolod1983 Will review tomorrow, sorry for delaying.

> `docs/readthedocs.com:advanced-micro-devices-miopen — Read the Docs build failed!`

Please check this. Changes in miopen.h may lead to such failures.

File: src/softmax.cpp [hunk 1]
```diff
@@ -33,8 +33,55 @@
...
 
+#include <nlohmann/json.hpp>
+
 namespace miopen {
 
+extern "C" miopenStatus_t miopenCreateSoftmaxDescriptor(miopenSoftmaxDescriptor_t* activDesc)
+{
+
+    MIOPEN_LOG_FUNCTION(activDesc);
+    return miopen::try_([&] { miopen::deref(activDesc) = new miopen::SoftmaxDescriptor(); });
```
Comment (DrizztDoUrden) [inline src/softmax.cpp]:
```suggestion
    return miopen::try_([&] {
        auto& desc = miopen::deref(activDesc);
        desc = new miopen::SoftmaxDescriptor();
    });
```
This way memory would not leak if deref throws

**Review by DrizztDoUrden — APPROVED** [review]
LGTM with one minor suggestion

File: include/miopen/miopen.h [hunk 2]
```diff
@@ -5314,6 +5319,11 @@ typedef enum
     miopenTensorBiasY        = 9,
     miopenTensorBias         = 10,
 #endif
+    miopenTensorSoftmaxX  = 11,
+    miopenTensorSoftmaxY  = 12,
+    miopenTensorSoftmaxDX = 13,
+    miopenTensorSoftmaxDY = 14,
```
Comment (atamazov) [inline include/miopen/miopen.h]: [Recommendation] I highly recommend moving these new members above the BETA API stuff and renumber items. This way, non-BETA members will have stable values regardless of what is happening with BETA things.

File: src/problem.cpp [hunk 3]
```diff
@@ -431,6 +463,60 @@ std::vector<Solution> Problem::FindSolutionsImpl(Handle& handle,
...
+    std::vector<solver::softmax::SoftmaxSolver*> solvers;
+
+    solvers.push_back(&attnSoftmaxSolver);
+    solvers.push_back(&regularSoftmaxSolver);
+
+    for(auto solver : solvers)
+    {
+        if(!solver->IsApplicable(ctx, problem_description))
+        {
+            continue;
```
Comment (atamazov) [inline src/problem.cpp:494]: [R] It may worth logging non-applicable Solvers at Info2 level, like we do for convolution solvers.

File: src/softmax.cpp [hunk 4]
```diff
@@ -33,8 +33,57 @@
...
 
+#include <nlohmann/json.hpp>
+
 namespace miopen {
 
+extern "C" miopenStatus_t miopenCreateSoftmaxDescriptor(miopenSoftmaxDescriptor_t* activDesc)
+{
+    MIOPEN_LOG_FUNCTION(activDesc);
+    return miopen::try_([&] {
+        auto& desc = miopen::deref(activDesc);
```
Comment (atamazov) [inline src/softmax.cpp]: [R] `activDesc` seems inadequate in this context, right? 😉

File: src/problem.cpp [hunk 5]
```diff
@@ -277,6 +282,33 @@ activ::ProblemDescription Problem::AsActivation() const
...
+    const auto& softmax_desc = boost::get<SoftmaxDescriptor>(operator_descriptor);
+
+    float alpha = softmax_desc.GetAlpha();
+    float beta  = softmax_desc.GetBeta();
+
+    softmax::ProblemDescription problem_description =
+        (GetDirection() == miopenProblemDirectionForward)
+            ? softmax::ProblemDescription(
+                  &alpha,
+                  &beta,
```
Comment (atamazov) [inline src/problem.cpp:296]: [Performance] @Vsevolod1983 See https://github.com/ROCm/MIOpen/pull/2671#pullrequestreview-1937760018

File: src/problem.cpp [hunk 6]
```diff
@@ -431,6 +463,60 @@ std::vector<Solution> Problem::FindSolutionsImpl(Handle& handle,
...
+
+    const auto algo = AlgorithmName{"Softmax"};
+
+    solver::softmax::AttnSoftmax attnSoftmaxSolver;
+    solver::softmax::Softmax regularSoftmaxSolver;
+
+    std::vector<solver::softmax::SoftmaxSolver*> solvers;
+
+    solvers.push_back(&attnSoftmaxSolver);
+    solvers.push_back(&regularSoftmaxSolver);
```
Comment (atamazov) [inline src/problem.cpp]: [Can be postponed] I think we can statically initialize `solvers`.

File: src/problem.cpp [hunk 7]
```diff
@@ -431,6 +463,60 @@ std::vector<Solution> Problem::FindSolutionsImpl(Handle& handle,
...
+        if(!solver->IsApplicable(ctx, problem_description))
+        {
+            continue;
+        }
+
+        auto solution = Solution();
+
+        /// \todo time measurement will be done later. For now we set less time for attention
+        /// softmax and slightly bigger for regular
+        solution.SetTime(solver == &attnSoftmaxSolver ? 0.0f : 1.0f);
```
Comment (atamazov) [inline src/problem.cpp]: No zero time, please.

[Q] Are we really going to benchmark the applicable Solutions? Or just use some kind of heuristics like WTI computations, neural nets or so?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: include/miopen/miopen.h (hunk 2 above)
Comment (Vsevolod1983) [reply include/miopen/miopen.h]: Renumbering existing items in enums are usually dangerous in case if they get serialized somewhere.
Are we sure they are not serialized ?

File: src/problem.cpp (hunk 7 above)
Comment (Vsevolod1983) [reply src/problem.cpp]: What do you suggest instead of zero time? 0.1f ?
As we told with @CAHEK7  probably later we add some time management, not a highly priority task though.

Comment (atamazov) [reply src/problem.cpp]: @Vsevolod1983 `? 1.0f : 2.0f);`

File: include/miopen/miopen.h (hunk 2 above)
Comment (atamazov) [reply include/miopen/miopen.h]: @Vsevolod1983 We do not care about BETA_API stuff. At least it has lower importance than "normal" API things, right?

Comment (Vsevolod1983) [reply include/miopen/miopen.h]: ok

File: src/problem.cpp (hunk 5 above)
Comment (Vsevolod1983) [reply src/problem.cpp:296]: I removed alpha and beta from network config in softmax primitive. Is it enough for the this PR ?

File: src/problem.cpp (hunk 6 above)
Comment (atamazov) [reply src/problem.cpp]: @DrizztDoUrden Can you please suggest @Vsevolod1983 the technique for doing this? Thanks.

File: src/softmax.cpp (hunk 4 above)
Comment (atamazov) [reply src/softmax.cpp]: [resolved]

File: src/problem.cpp (hunk 7 above)
Comment (atamazov) [reply src/problem.cpp]: [resolved]

File: src/problem.cpp [hunk 8]
```diff
@@ -490,14 +490,15 @@ Problem::FindSolutionsImpl(Handle& handle,
     {
         if(!solver->IsApplicable(ctx, problem_description))
         {
+            MIOPEN_LOG_I2("solver " << solver->SolverDbId() << "is not applicable for the given problem description");
```
Comment (atamazov) [inline src/problem.cpp]: Space before "is" is missing. Please borrow message format from there: https://github.com/ROCm/MIOpen/blob/cee33a7aaa3dbbe850d2580a72fff8774ddbc131/src/include/miopen/find_solution.hpp#L209

File: include/miopen/miopen.h [hunk 9]
```diff
@@ -5310,19 +5310,20 @@ typedef enum
     miopenTensorConvolutionX      = 1,
     miopenTensorConvolutionW      = 2,
     miopenTensorConvolutionY      = 3,
+    miopenTensorSoftmaxX  = 4,
```
Comment (atamazov) [inline include/miopen/miopen.h]: [Informative] I suspect that clang-format will fail. See https://github.com/ROCm/MIOpen/wiki/How-to-format-code

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/problem.cpp (hunk 5 above)
Comment (atamazov) [reply src/problem.cpp:296]: @Vsevolod1983 I would begin with removing alpha/beta from PD (and then fix build errors by forwarding alpha/beta via InvokeParams). However I think we need to discuss https://github.com/ROCm/MIOpen/pull/2671#pullrequestreview-1937760018 first.

/cc @CAHEK7 @DrizztDoUrden

File: include/miopen/miopen.h (hunk 2 above)
Comment (atamazov) [reply include/miopen/miopen.h]: [resolved]

File: src/problem.cpp (hunk 3 above)
Comment (atamazov) [reply src/problem.cpp:494]: [resolved]

**Comment by atamazov** [comment]
@Vsevolod1983 Please hide all resolved review comments to save space (my review comments are resolved when marked with `[resolved]`), thanks!

File: src/problem.cpp (hunk 5 above)
Comment (Vsevolod1983) [reply src/problem.cpp:296]: Should we do this kind of refactoring as a separate ticket / PR ?

Comment (CAHEK7) [reply src/problem.cpp:296]: We need. Actually a/b affects kernel compilation, like it does for convolutions. There are some optimizations with default a/b values, so we need at least a "default a/b" flag in the problem description and in the network config.

Comment (Vsevolod1983) [reply src/problem.cpp:296]: I restored alpha and beta in network config in this PR until we decide what exactly we want to do next.

Comment (atamazov) [reply src/problem.cpp:296]: [Informative] @Vsevolod1983 @CAHEK7 it's interesting that "default alpha" optimization won't give us much, but "default beta" opt removes 1 global memory read, which may be substantial (./src/kernels/MIOpenSoftmax.cl)

File: src/problem.cpp (hunk 8 above)
Comment (atamazov) [reply src/problem.cpp]: [resolved]

**Review by atamazov — APPROVED** [review]
🟢 LGTM!

Postponed topics:
- https://github.com/ROCm/MIOpen/pull/2776#discussion_r1525541568 - alpha/beta thing
- https://github.com/ROCm/MIOpen/pull/2776#discussion_r1525544719 - static init of vector of solvers

File: src/problem.cpp (hunk 5 above)
Comment (CAHEK7) [reply src/problem.cpp:296]: @atamazov right now `default alpha` gives much since it enables attention softmax solver, which is faster and does not support a/b in terms of normal softmax operation.

Both values are used in the `isApplicable` method and can't be just removed.

**Review by junliume — APPROVED** [review]

File: src/problem.cpp (hunk 6 above)
Comment (DrizztDoUrden) [reply src/problem.cpp]: The most basic primitive for this is https://github.com/ROCm/MIOpen/blob/9db0c265deb57b33d1a5669264157bc2700bd6d8/src/include/miopen/find_solution.hpp#L181

File: src/include/miopen/softmax.hpp [hunk 10]
```diff
@@ -28,12 +28,47 @@
...
+    }
+
+    friend std::ostream& operator<<(std::ostream& stream, const SoftmaxDescriptor& x);
+
+    friend void to_json(nlohmann::json& json, const SoftmaxDescriptor& descriptor);
+    friend void from_json(const nlohmann::json& json, SoftmaxDescriptor& descriptor);
+
+private:
+    float alpha;
+    float beta;
```
Comment (DrizztDoUrden) [inline src/include/miopen/softmax.hpp:67]: I would rather store them as `bool`. This way later it would be harder to accidentally use them instead of actual values. On subsequent runs float values here would be incorrect anyway.
```suggestion
    bool has_alpha;
    bool has_beta;
```

File: src/solution.cpp [hunk 11]
```diff
@@ -232,6 +241,91 @@ void Solution::RunImpl(Handle& handle,
...
+        }
+        auto ret = found->second;
+        if(!ret.descriptor.has_value())
+            ret.descriptor = problem_casted.GetTensorDescriptorChecked(name, name_str);
+        return ret;
+    };
+
+    const softmax::ProblemDescription problem_description = problem_casted.AsSoftmax();
+
+    float alpha                        = softmax_desc.GetAlpha();
```
Comment (DrizztDoUrden) [inline src/solution.cpp:268]: You can't use these values, they are outdated on subsequent calls. User should be able to invoke the same solution multiple times with different a/b, unless they were defaulted in find.

**Review by DrizztDoUrden — CHANGES_REQUESTED** [review]

**Comment by DrizztDoUrden** [comment]
If we merge it with a/b set through PD, it would be kind of a breaking change later. To do so it has to be wrapped in `MIOPEN_BETA_API`

File: src/problem.cpp (hunk 6 above)
Comment (DrizztDoUrden) [reply src/problem.cpp]: This is postponable in contrast to a/b problem

**Comment by DrizztDoUrden** [comment]
I have somewhat hacky solution for passing scalar arguments at find and run time here: https://github.com/ROCm/MIOpen/pull/2825

**Comment by DrizztDoUrden** [comment]
I'll split scalar argument passing to a separate [PR](https://github.com/ROCm/MIOpen/pull/2826) today

**Comment by Vsevolod1983** [comment]
I will use MIOPEN_BETA_API for this PR to have it merged. When we get understanding on how to refactor, lets do it in a separate ticket

**Comment by Vsevolod1983** [comment]
@DrizztDoUrden 
I put miopenSoftmaxCreateProblem declaration under MIOPEN_BETA_API. Tensor's enum values as well.

Btw, just an observation. Our library can be built only with this beta api flag. 
We have   add_definitions("-DMIOPEN_BETA_API=1") in cmakelist.txt file.
Without this flag library cannot be compiled.

**Review by DrizztDoUrden — APPROVED** [review]
LGTM
